### DISCUSSION TITLE
Functions and Objects

### DISCUSSION THREAD

**Discussion by ebarrragn** [Maximilian-Winter/llama-cpp-agent#33] [Q&A] [answered]
Hi, 
I am trying to figure out how to call an object's method from a callable function, I have tried different approaches valid in python, but they end in Pydantic errors. 
Basically, I am playing with agents spawning other agents as needed, but for that I need some way to retrieve the caller (an object) from the callable function, so it can spawn a child agent...
Any suggestions are welcome.

**Comment by Maximilian-Winter** [comment, marked as answer]
@ebarrragn You can pass additional key word arguments to the LlamaCppFunctionTool. They will be passed to the run function.

**Reply by ebarrragn** [reply]
I can see it now in the '__call__' method, I missed it, I'm not that good at Python.
Thanks!

**Comment by Maximilian-Winter** [comment]
@ebarrragn If you need help with that, let me know.

**Comment by ebarrragn** [comment]
Solved, read at the end.

Thanks, Maximilian, I tried, but I can't figure it out.
For instance, in the file I am attaching, `ListChildrenAgents` needs access to the caller `AgentWithChildren` instance to list the children it has, using the `AgentWithChildren.children_descriptions() `method. I am trying to save context space, so this information is not injected at every prompt, otherwise, I could have done it statically at object creation.

You can see other attempts to save context space when I save/restore the llama_cpp state in `AgentContainer.run_child_agent()`.
The agent calls `ListChildrenAgents` every time it needs to schedule a call to one of its children.

The need of having the caller also appears in `SpawnChildAgent`, it needs access to the parent to spawn one of its children.

I suspect I am doing something fundamentally wrong.

[tkagents.py.txt](https://github.com/Maximilian-Winter/llama-cpp-agent/files/15370519/tkagents.py.txt)

===========================================================================================================

Well, it works, Just explaining how I did it:

I am creating each tool, inside a container object, as:

`lcft = LlamaCppFunctionTool(tool, caller_agent=self)  # tool is a BaseModel derived class` 
for each tool I changed the run method to be:
`def run(self, caller_agent):`

I had some troubles (exceptions about extra parameters) but they disappeared once I relaunched the environment, I suppose I forgot to save a file after making the changes.
